### PR TITLE
Add support for Container image digest reference

### LINKED ISSUE #4169: Container image repository name validation when using image digest instead of tag
When the container image of a stream/task application is registered with SCDF, it could have image digest or tag. For instance,

I can have an app like follows:

`docker:mydevapps.io/p-scdf-for-kubernetes/spring-cloud-dataflow-composed-task-runner:<tag>`

`docker:mydevapps.io/p-scdf-for-kubernetes/spring-cloud-dataflow-composed-task-runner@sha256:<hash>`

When the image metadata is retrieved, the container image's repository name is currently parsed with the assumption of having a `tag` in the image name and this breaks when I have an app with the image digest in it.

### PR DESCRIPTION
**Pull request by tzolov**
- Extend the ContainerImage and ContainerImageParser to allow image digest along with existing image tags.

 Resolves #4169

### CODE REVIEW

File: spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/ContainerImage.java [hunk 1]
```diff
@@ -55,10 +55,13 @@
 	// and dashes. A tag name may not start with a period or a dash and may contain a maximum of 128 characters.
 	// (https://dockr.ly/3chhQZF)
 	private static final Pattern TAG_PATTERN = Pattern.compile("^[a-zA-Z0-9_][a-zA-Z0-9\\-_.]{0,127}$");
+	private static final Pattern DIGEST_PATTERN = Pattern.compile("^[A-Za-z][A-Za-z0-9]*(?:[\\-_+.][A-Za-z][A-Za-z0-9]*)*[:][[\\p{XDigit}]]{32,}$");
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/ContainerImage.java:58]: Is this from the standard format (specified somewhere?) for digest similar to tag pattern above?

File: spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/ContainerImage.java [hunk 2]
```diff
@@ -155,19 +162,52 @@ public void setRepositoryName(String repositoryName) {
...
 	public String getRepositoryTag() {
 		return repositoryTag;
 	}
 
 	public void setRepositoryTag(String repositoryTag) {
-		Assert.isTrue(TAG_PATTERN.matcher(repositoryTag).matches(),
-				"Invalid repository tag: " + repositoryTag);
+		Assert.isTrue(TAG_PATTERN.matcher(repositoryTag).matches(), "Invalid repository tag: " + repositoryTag);
+		Assert.isTrue(!StringUtils.hasText(this.repositoryDigest),
+				"Can set repository Tag because of existing Digest " + repositoryDigest);
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/ContainerImage.java:192]: you mean `Can't set repository Tag` right?

File: spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/ContainerImage.java [hunk 3]
```diff
@@ -155,19 +162,52 @@ public void setRepositoryName(String repositoryName) {
...
 	}
 
+	public String getRepositoryDigest() {
+		return repositoryDigest;
+	}
+
+	public void setRepositoryDigest(String repositoryDigest) {
+		Assert.isTrue(DIGEST_PATTERN.matcher(repositoryDigest).matches(), "Invalid repository digest: " + repositoryDigest);
+		Assert.isTrue(!StringUtils.hasText(this.repositoryTag),
+				"Can set repository digest because of existing tag " + repositoryTag);
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/ContainerImage.java:203]: Can't set ?

File: spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/ContainerImage.java (hunk 1 above)
Comment (tzolov) [reply spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/ContainerImage.java:58]: This from the docker's own source code : https://github.com/docker/distribution/blob/master/reference/regexp.go#L44

File: spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/ContainerImage.java (hunk 2 above)
Comment (tzolov) [reply spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/ContainerImage.java:192]: It is `Can't` indeed

File: spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/ContainerImage.java (hunk 3 above)
Comment (tzolov) [reply spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/ContainerImage.java:203]: Same it is "Can't"

Comment (ilayaperumalg) [reply spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/ContainerImage.java:203]: ok, will fix the changes and merge.

**Comment by ilayaperumalg** [comment]
LGTM, merging.

**Comment by ilayaperumalg** [comment]
Merged as 415dee03f

**Comment by ilayaperumalg** [comment]
Cherry-picked the changes and pushed into 2.6.x as a7480171185d7e68072e135bd71fc4b579e8d2c2

**Pull request closed without merging**